### ISSUE TITLE
[BUG] `T.gemm` outputs NaN when kernel has multiple thread blocks

### ISSUE THREAD

**Issue by One-sixth** [tile-ai/tilelang#2936]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.13 (also reproduced on 0.1.12)

### System information

- TileLang version: 0.1.13 (also reproduced on 0.1.12)
- GPU: NVIDIA RTX 4060 Laptop (Ada Lovelace, Driver 610.88)
- CUDA: 13.0
- OS: Windows 11
- Python: 3.13

## Description

`T.gemm` produces incorrect results (all NaN / garbage values) when the `T.Kernel` grid has more than 1 block in any dimension. The same `T.gemm` call works correctly with a single-block grid `(1, 1)`.

## Expected Behavior

`T.gemm` should produce the same result regardless of the number of thread blocks. Multi-block execution should give correct results (within float32 precision, max_diff ~0.016).

## Actual Behavior

With `T.Kernel(1, 16, threads=64)`, all `T.gemm` outputs are incorrect — either NaN, extremely large garbage values (~1e38), or wrong values. With `T.Kernel(1, 1, threads=64)`, the same computation is correct.

The kernel does execute and write to the output tensor (output values differ from initialization), but the computed values are garbage.

## Additional Observations

- The bug occurs with `threads=64` (2 warps). Also tested with `threads=128` (4 warps) — same result.
- Matrix size: `[16, 64] @ [64, 16] = [16, 16]`
- `T.Parallel` and `T.copy` work correctly in multi-block mode; only `T.gemm` is affected.
- The kernel **does execute** (output is not the initial value), but `T.gemm` produces wrong results.
- **First call returns NaN, subsequent calls return stable-but-wrong values** (e.g. ~2.3e33). This suggests uninitialized shared memory or register state within `T.gemm`'s internal implementation — the first execution reads garbage (NaN), subsequent executions read the same garbage from the previous run (stable but wrong).
- The issue is likely within `T.gemm`'s internal workspace initialization, not the user-facing shared memory buffers.

## Workaround

Replacing `T.gemm` with manual matrix multiplication using `T.Parallel` + sequential loops works correctly in multi-block mode:

```python
for i, j in T.Parallel(BC, BC):
    s = T.alloc_fragment((1,), torch.float32)
    s[0] = 0.0
    for kk in range(BK):
        s[0] = s[0] + a_s[i, kk] * b_s[kk, j]
    c_f[i, j] = s[0]
```

## Minimal Reproduction

```python
import torch
import tilelang
import tilelang.language as T

BK, BC = 64, 16

@tilelang.jit
def gemm_single_block():
    @T.prim_func
    def kernel(
        A: T.Tensor((BC, BK), torch.float32),
        B: T.Tensor((BK, BC), torch.float32),
        C: T.Tensor((BC, BC), torch.float32),
    ):
        with T.Kernel(1, 1, threads=64) as (bx, by):
            a_s = T.alloc_shared((BC, BK), torch.float32)
            b_s = T.alloc_shared((BK, BC), torch.float32)
            T.copy(A, a_s)
            T.copy(B, b_s)
            c_f = T.alloc_fragment((BC, BC), torch.float32)
            T.gemm(a_s, b_s, c_f)
            T.copy(c_f, C)
    return kernel

@tilelang.jit
def gemm_multi_block(nblocks=16):
    @T.prim_func
    def kernel(
        A: T.Tensor((nblocks, BC, BK), torch.float32),
        B: T.Tensor((nblocks, BK, BC), torch.float32),
        C: T.Tensor((nblocks, BC, BC), torch.float32),
    ):
        with T.Kernel(1, nblocks, threads=64) as (bx, bn):
            a_s = T.alloc_shared((BC, BK), torch.float32)
            b_s = T.alloc_shared((BK, BC), torch.float32)
            T.copy(A[bn, :, :], a_s)
            T.copy(B[bn, :, :], b_s)
            c_f = T.alloc_fragment((BC, BC), torch.float32)
            T.gemm(a_s, b_s, c_f)
            T.copy(c_f, C[bn, :, :])
    return kernel

# Test single block -- works correctly
A_single = torch.randn(BC, BK, device='cuda', dtype=torch.float32)
B_single = torch.randn(BK, BC, device='cuda', dtype=torch.float32)
C_single = torch.zeros(BC, BC, device='cuda', dtype=torch.float32)
fn1 = gemm_single_block()
fn1(A_single, B_single, C_single)
ref = A_single @ B_single
print(f"Single block: max_diff = {(C_single - ref).abs().max().item():.6f}")
# Output: max_diff = 0.016115 (correct, within float32 precision)

# Test multi block -- ALL NaN!
A_multi = torch.randn(16, BC, BK, device='cuda', dtype=torch.float32)
B_multi = torch.randn(16, BK, BC, device='cuda', dtype=torch.float32)
C_multi = torch.zeros(16, BC, BC, device='cuda', dtype=torch.float32)
fn2 = gemm_multi_block(nblocks=16)
fn2(A_multi, B_multi, C_multi)
print(f"Multi block: max_diff = {(C_multi - (A_multi @ B_multi)).abs().max().item()}")
# Output: nan (INCORRECT!)
```

### Traceback

```pytb

```

**Comment by RyanL2** [comment]
This looks like a missing accumulator initialization in the reproducer rather than a multi-block lowering bug.

`T.gemm` **accumulates** into its `C` operand — it does not overwrite it. The signature in 0.1.13 is:

```
T.gemm(A, B, C, transpose_A=False, transpose_B=False,
       policy=GemmWarpPolicy.Square, clear_accum=False, k_pack=1, mbar=None)
```

Since `clear_accum` defaults to `False`, `T.gemm` adds into whatever those registers already held. In the reproducer `c_f` is never initialized:

```python
c_f = T.alloc_fragment((BC, BC), torch.float32)
T.gemm(a_s, b_s, c_f)          # c_f holds whatever was in those registers
```

The fix is one line:

```python
c_f = T.alloc_fragment((BC, BC), torch.float32)
T.clear(c_f)                   # or: T.gemm(a_s, b_s, c_f, clear_accum=True)
T.gemm(a_s, b_s, c_f)
```

This matches the convention everywhere else in the repo — `examples/gemm/example_gemm.py` calls `T.clear(C_local)` before its `T.gemm`, and across `examples/` the `clear_accum` argument appears 101 times in 40 files, with `T.clear(` appearing 346 times.

### The generated CUDA

This part doesn't depend on hardware. Dumping `get_kernel_source()` for the reproducer, `c_f` appears exactly three times in the whole kernel:

```c
  float c_f[4];                                          // declared, no initializer
  ...
    for (int ki = 0; ki < 8; ++ki) {
      ...
      tl::mma_sync<...>(reinterpret_cast<float*>(c_f + 0),
                        reinterpret_cast<const unsigned*>(A_local + 0),
                        reinterpret_cast<const unsigned*>(B_local + 0));   // reads + accumulates
    }
  ...
    *(float2*)(C + ...) = *(float2*)(c_f + (i_2 * 2));    // copied out
```

At `ki == 0` the MMA reads `c_f` before anything has written to it.

Dumping the same kernel with `T.clear(c_f)` added and diffing the two, the entire difference is two added lines:

```diff
+  float broadcast_var = 0x0p+0f/*0.000000e+00*/;
+  *(float4*)(c_f + 0) = make_float4(broadcast_var, broadcast_var, broadcast_var, broadcast_var);
```

Nothing else changes — same addressing, same `__syncthreads()`, same MMA instructions. Zeroing the accumulator is the only thing `T.clear` contributes.

### Why this explains the observations

| Observation | Explanation |
| --- | --- |
| Single block works, 16 blocks doesn't | Whether the stale registers happen to be zero depends on what ran before, not on the block count. The single-block case wasn't correct, it was lucky. |
| 64 and 128 threads both fail | Thread count isn't the variable. |
| `T.copy` / `T.Parallel` are fine | They write their destination; they don't accumulate into it. |
| The manual matmul works | The workaround does `s[0] = 0.0` — it initializes the accumulator. |
| First call NaN, later calls stable garbage | Uninitialized register reuse across launches. |

The manual-matmul workaround is the strongest tell here: it already contains the fix.

### What I measured, and what I didn't

I'm on an **RTX 3060 Ti (sm_86)**, not your RTX 4060 Laptop (sm_89), running TileLang 0.1.13.

**I did not reproduce the NaN.** On my GPU the uninitialized version returns *correct* results in every configuration I tried. I don't think that argues against the explanation — it's what reading uninitialized registers looks like. The value is undefined, so whether it happens to be zero varies with architecture, driver, and whatever ran before. The same program being correct on sm_86 and NaN on sm_89 is the symptom, not a counterexample.

With the accumulator explicitly cleared, every configuration I swept was correct (max_diff < 0.05 against `torch.bmm`): grids `(1,1)`, `(1,2)`, `(1,4)`, `(1,16)`, `(2,1)`, `(4,1)` at both 64 threads (16×16 tile) and 128 threads (32×32 tile — a 16×16 output tile can't be partitioned across 4 warps, which `T.gemm` rejects at compile time with a clear message).

I ran that sweep specifically to try to *falsify* this explanation: if a cleared accumulator still failed at multi-block, there would be a real lowering bug underneath and the initialization would just be masking it. It didn't fail.

### Conclusion

I couldn't reproduce the NaN symptom, but the read-before-initialization is deterministic in the generated code, the API defaults to accumulation, and explicitly initializing the accumulator makes every configuration I tested correct. On that evidence I don't think this is a multi-block lowering bug.

If it still fails with `T.clear(c_f)` on the original machine, that would be evidence of an additional bug — but the uninitialized accumulator accounts for everything reported here on its own.

**Comment by RyanL2** [comment]
Opened #2939 with a compile-time check for this: it warns when a `T.gemm` accumulator can't be shown to be initialized before use, and stays quiet on the pipelined `clear_accum=(k == 0)` pattern. No warnings across the in-tree GEMM, transform, and language test suites.

**Comment by One-sixth** [comment]
@RyanL2 Thank you — the uninitialized accumulator was indeed the root cause. I've confirmed the fix and migrated the remaining manual matmuls to T.gemm with clear_accum=True. Thanks!
